File: q2-backend/ts_template/roundupper_100.ts
import express from 'express';
import bodyParser from 'body-parser';

// location is the simple (x, y) coordinates of an entity within the system
// spaceCowboy models a cowboy in our super amazing system
// spaceAnimal models a single animal in our amazing system
type location = { x: number, y: number };
type spaceCowboy = { lassoLength: number, location: location };
type spaceAnimal = { type: "pig" | "cow" | "flying_burger", location: location };


// === ADD YOUR CODE BELOW :D ===

// === ExpressJS setup + Server setup ===
const spaceAnimals = [] as spaceAnimal[];
const spaceCowboys: { [name: string]: spaceCowboy } = {};
const app = express();
app.use(bodyParser.json());

// the POST /entity endpoint adds an entity to your global space database
app.post('/entity', (req, res) => {
  const entities = req.body["entities"];

  for (const entity of entities) {
    const metadata = entity["metadata"];
    const location: location = entity["location"];

    switch (entity["type"]) {
      case "space_cowboy":
        spaceCowboys[metadata["name"]] = { lassoLength: metadata["lassoLength"], location };
        break;

      case "space_animal":
        const type = metadata["type"];
        if (!["pig", "cow", "flying_burger"].includes(type)) {
          res.status(400).send("Invalid Space Animal");
          return;
        }
        spaceAnimals.push({ type, location });
        break;

      default:
        res.status(400).send("Invalid Entity Type");
        return;
    }
  }

  res.status(200).send("OK");
});

// lassoable returns all the space animals a space cowboy can lasso given their name
app.get('/lassoable', (req, res) => {
  const name = req.body["cowboy_name"];

  const cowboy = spaceCowboys[name];
  const space_animals = spaceAnimals.filter(animal => {
    const distance = Math.sqrt(Math.pow(animal.location.x - cowboy.location.x, 2) + Math.pow(animal.location.y - cowboy.location.y, 2));
    return distance <= cowboy.lassoLength;
  });

  res.status(200).send({ space_animals })
});

app.listen(8080);

File: q2-backend/py_template/roundupper_100.py
from dataclasses import dataclass
from enum import Enum
from typing import NamedTuple
from flask import Flask, request
from math import dist

import sys

# Please update to Python 3.11
if sys.version_info >= (3, 11):
    from typing import Self
else:
    from typing_extensions import Self

# Location models a location in our amazing system
@dataclass
class Location:
    x: int
    y: int

    def distance_to(self, other: Self) -> float:
        return dist((self.x, self.y), (other.x, other.y))


# SpaceCowboy models a cowboy in our super amazing system
@dataclass
class SpaceCowboy:
    lasso_length: int
    location: Location


# SpaceAnimalType is an enum of all possible space animals we may encounter
class SpaceAnimalType(str, Enum):
    PIG = "pig"
    COW = "cow"
    FLYING_BURGER = "flying_burger"


# SpaceAnimal models a single animal in our amazing system
@dataclass
class SpaceAnimal:
    type: SpaceAnimalType
    location: Location


# ==== HTTP Endpoint Stubs ====
app = Flask(__name__)
space_animals: list[SpaceAnimal] = []
space_cowboys: dict[str, SpaceCowboy] = {}

# the POST /entity endpoint adds an entity to your global space database
@app.route("/entity", methods=["POST"])
def create_entity():
    entities = request.get_json()["entities"]

    for e in entities:
        meta = e["metadata"]
        location = Location(**e["location"])

        match e["type"]:
            case "space_cowboy":
                # Not handling duplicate cowboy names as guaranteed to be unique
                space_cowboys[meta["name"]] = SpaceCowboy(meta["lassoLength"], location)

            case "space_animal":
                match (meta["type"]):
                    case "pig":
                        type = SpaceAnimalType.PIG
                    case "cow":
                        type = SpaceAnimalType.COW
                    case "flying_burger":
                        type = SpaceAnimalType.FLYING_BURGER
                    case _:
                        return "Invalid animal type", 400
                space_animals.append(SpaceAnimal(type, location))

            case _:
                return "Invalid entity type", 400

    return "OK", 200


# lassoable returns all the space animals a space cowboy can lasso given their name
@app.route("/lassoable", methods=["GET"])
def lassoable():
    name = request.get_json()["cowboy_name"]
    cowboy = space_cowboys[name]
    lassoable = [
        animal
        for animal in space_animals
        if cowboy.location.distance_to(animal.location) <= cowboy.lasso_length
    ]

    return {"space_animals": lassoable}, 200


# DO NOT TOUCH ME, thanks :D
if __name__ == "__main__":
    app.run(debug=True, port=8080)
